src/transformStateWithClones.js

```diff
@@ -6,8 +6,30 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  let newState = { ...state };
+
+  for (const obj of actions) {
+    if (obj.type === 'addProperties') {
+      newState = { ...newState, ...obj.extraData };
+
+      result.push(newState);
+    } else if (obj.type === 'removeProperties') {
+      for (const key of obj.keysToRemove) {
+        const { [key]: removed, ...rest } = newState;
+
+        newState = rest;
+      }
+      result.push({ ...newState });
+    } else if (obj.type === 'clear') {
+      newState = {};
+      result.push({ ...newState });
```

Comment:
According to the checklist, it is better to push a copy of the object to the array at the end of each loop cycle, but outside of the switch block, after the current action is processed.